### PR TITLE
Add Decoder for The Graph staking

### PR DESCRIPTION
**Pull request by d347h-eth**
Related to #(2044)

For Delegator staking the typical scenario is:
- call `delegate()` (emits `StakeDelegated` event)
- call `undelegate()` (emits `StakeDelegatedLocked` event)
- call `withdrawDelegated()` (emits `StakeDelegatedWithdrawn` event)

### CODE REVIEW

File: rotkehlchen/chain/ethereum/modules/the_graph/constants.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+
+CPT_THE_GRAPH = 'the_graph'
+
+THE_GRAPH_CPT_DETAILS = CounterpartyDetails(
+    identifier=CPT_THE_GRAPH,
+    label='The Graph',
+    image='',  # TODO: add image
```
Comment (d347h-eth) [inline rotkehlchen/chain/ethereum/modules/the_graph/constants.py:8]: how to add this?

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            tx_log=context.tx_log,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=A_GRT,
+            balance=Balance(),
+            location_label=delegator,
+            notes=(
+                f'Undelegated {tokens_amount_norm} GRT'
+                f' from Indexer {indexer}'
+                f', lock expires in {lock_timeout_secs} seconds'
```
Comment (d347h-eth) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:143]: The reason I didn't use `,` in the end of line: https://github.com/rotki/rotki/issues/5432#issuecomment-1750724659

File: rotkehlchen/chain/ethereum/modules/the_graph/constants.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/the_graph/constants.py:8]: we explained this in discord. Ask again if it does not make sense

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            },
+        }}
+
+    def addresses_to_decoders(self) -> dict[ChecksumEvmAddress, tuple[Any, ...]]:
+        return {
+            CONTRACT_THE_GRAPH_STAKING: (self._decode_delegator_staking,),
+        }
+
+    # processes events from The Graph's staking contract
+    def _decode_delegator_staking(self, context: DecoderContext) -> DecodingOutput:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:63]: Avoid simple comments when it's meant for functions or methods. Turn into docstring

```suggestion
    def _decode_delegator_staking(self, context: DecoderContext) -> DecodingOutput:
         """Processes events from The Graph's staking contract"""
```

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+from rotkehlchen.types import ChecksumEvmAddress, DecoderEventMappingType
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+if TYPE_CHECKING:
+    from rotkehlchen.assets.asset import EvmToken
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+CONTRACT_THE_GRAPH_STAKING = string_to_evm_address('0xF55041E37E12cD407ad00CE2910B8269B01263b9')
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:32]:
```suggestion
CONTRACT_THEGRAPH_STAKING = string_to_evm_address('0xF55041E37E12cD407ad00CE2910B8269B01263b9')
```

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        )
+        # find Transfer event with delegation tax (0.5% GRT burn) and persist the burnt amount
+        tokens_burnt = 0
+        for ctx_log in context.all_logs:
+            if (
+                ctx_log.topics[0] == TOPIC_TRANSFER and
+                hex_or_bytes_to_address(ctx_log.topics[1]) == CONTRACT_THE_GRAPH_STAKING and
+                hex_or_bytes_to_address(ctx_log.topics[2]) == ZERO_ADDRESS
+            ):
+                tokens_burnt = hex_or_bytes_to_int(ctx_log.data)
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:95]: This is good and will work for the example you have. The problem will come if someone writes a contract/script that does multiple delegations in 1 transaction.

Then this approach will find the last one. Notice you don't stop (do a `break`) in the for loop once you find it, thus you find the last burnt event.

Since you mentioned in discord that the burn % is fixed, then perhaps we can use this to our advantage in order to be even more accurate.

So find the burned event that matches the 0.5% of the staked amount, use that and then break out of the searching for loop

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        # identify and override the original stake Transfer event
+        for event in context.decoded_events:
+            if (
+                event.location_label == delegator and
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE
+            ):
+                event.event_type = HistoryEventType.STAKING
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.notes = f'Delegated {stake_amount} GRT to Indexer {indexer}'
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:105]: hmmm no no. You actually need a separate event for the burning. You need to use the burning event. And create a spend/fee or something like that event. Otherwise accounting will be off. You send 1000 GRT, and only 995 arrive. The rest are lost and not able to account for. 

Remember `notes` are human readable, but completely ignored when processing events. Events are also supposed to be for 1 asset movement either in or out. And the way you currently do it, we are missing 0.05% of the tokens each time

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                if tokens_burnt > 0:
+                    taxed_amount = token_normalized_value(
+                        token_amount=tokens_burnt,
+                        token=token,
+                    )
+                    event.notes += f', {taxed_amount} GRT has been burnt as delegation tax'
+                event.counterparty = CPT_THE_GRAPH
+        return DEFAULT_DECODING_OUTPUT
+
+    # processes StakeDelegatedLocked event
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:115]: same as before here

```suggestion
```

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py (hunk 2 above)
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:143]: no worries. I explained why in discord. It's unrelated to that issue.

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                f'Undelegated {tokens_amount_norm} GRT'
+                f' from Indexer {indexer}'
+                f', lock expires in {lock_timeout_secs} seconds'
+            ),
+            counterparty=CPT_THE_GRAPH,
+            address=context.tx_log.address,
+        )
+        return DecodingOutput(event=event)
+
+    # processes StakeDelegatedWithdrawn event
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:150]: same here

```suggestion
```

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+            from_event_type=HistoryEventType.RECEIVE,
+            from_event_subtype=HistoryEventSubType.NONE,
+            asset=A_GRT,
+            amount=tokens_amount_norm,
+            to_event_type=HistoryEventType.STAKING,
+            to_event_subtype=HistoryEventSubType.REMOVE_ASSET,
+            to_notes=(
+                f'Withdrew {tokens_amount_norm} GRT'
+                f' from Indexer {indexer}'
+            ),
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:176]: Doesn' this fit in one line? If it's just for a tiny bit you can also add #noqa: E501
If they fit in one line keep them as such.

```suggestion
            to_notes=f'Withdrew {tokens_amount_norm} GRT from Indexer {indexer}',
```

File: rotkehlchen/tests/unit/decoders/test_the_graph.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
+import pytest
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_the_graph.py:1]: same for renaming this file

File: rotkehlchen/tests/unit/decoders/test_the_graph.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+from rotkehlchen.chain.ethereum.modules.the_graph.constants import CPT_THE_GRAPH
+from rotkehlchen.chain.evm.decoding.constants import CPT_GAS
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.constants.assets import A_ETH, A_GRT
+from rotkehlchen.fval import FVal
+from rotkehlchen.tests.utils.ethereum import get_decoded_events_of_transaction
+from rotkehlchen.types import Location, TimestampMS, deserialize_evm_tx_hash
+
+ADDY_USER = '0xd200aeEC7Cd9dD27CAB5a85083953a734D4e84f0'
+ADDY_THE_GRAPH = '0xF55041E37E12cD407ad00CE2910B8269B01263b9'
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_the_graph.py:15]:
```suggestion
ADDY_THEGRAPH = string_to_evm_address('0xF55041E37E12cD407ad00CE2910B8269B01263b9')
```

File: rotkehlchen/tests/unit/decoders/test_the_graph.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.chain.ethereum.modules.the_graph.constants import CPT_THE_GRAPH
+from rotkehlchen.chain.evm.decoding.constants import CPT_GAS
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.constants.assets import A_ETH, A_GRT
+from rotkehlchen.fval import FVal
+from rotkehlchen.tests.utils.ethereum import get_decoded_events_of_transaction
+from rotkehlchen.types import Location, TimestampMS, deserialize_evm_tx_hash
+
+ADDY_USER = '0xd200aeEC7Cd9dD27CAB5a85083953a734D4e84f0'
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_the_graph.py:14]:
```suggestion
ADDY_USER = string_to_evm_address('0xd200aeEC7Cd9dD27CAB5a85083953a734D4e84f0')
```

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py (hunk 6 above)
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:105]: also lowercase indexer. No need to capitalize. Here and in other places

```suggestion
                event.notes = f'Delegated {stake_amount} GRT to indexer {indexer}'
```

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            transaction=context.transaction,
+            tx_log=context.tx_log,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=A_GRT,
+            balance=Balance(),
+            location_label=delegator,
+            notes=(
+                f'Undelegated {tokens_amount_norm} GRT'
+                f' from Indexer {indexer}'
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:142]:
```suggestion
                f' from indexer {indexer}'
```

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py (hunk 6 above)
Comment (d347h-eth) [reply rotkehlchen/chain/ethereum/modules/the_graph/decoder.py:105]: Let's validate I've understood this correctly. In the end we should have:
1) STAKING/DEPOSIT_ASSET event with balance=995 GRT
2) SPEND/FEE event with balance=5 GRT
Is that correct?

Comment (d347h-eth) [reply rotkehlchen/chain/ethereum/modules/the_graph/decoder.py]: Updated. There's informational event between STAKING/DEPOSIT_ASSET and [SPEND/FEE](https://github.com/rotki/rotki/pull/6712/files#diff-a44165e5a20fc6ff8b5f163bfcab946d05dd19e62835576c365aa941790408ccR75) now - I'm assuming that also should work?

File: rotkehlchen/chain/ethereum/modules/the_graph/decoder.py (hunk 5 above)
Comment (d347h-eth) [reply rotkehlchen/chain/ethereum/modules/the_graph/decoder.py]: [Updated](https://github.com/rotki/rotki/pull/6712/files#diff-95ffe39eb01c3fbcb9f750f20cf51d0e861a7be760f5e4921c42105b5e12e91dR102). I through through the possible variants on how to process it and figured that the subtraction is most okay-ish approach (and it also will work correct with multiple invocations of `StakeDelegated` decoding).
There's also not much reason to loop over `context.all_logs` and look for burn Transfers after all. Because the previous looping was through the logs - not decoded events accounting for burns. So we have nothing to re-purpose/override here and the search for the sake of validation doesn't have much meaning. We just need to add new events that will account for delegation taxes. So we just act on the assumption that every GRT transfer to the contract (that we didn't re-purpose/override yet) will create the relevant delegation tax spending event.

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+                initial_amount_norm = event.balance.amount
+                event.event_type = HistoryEventType.STAKING
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.balance = Balance(amount=stake_amount_norm)
+                event.notes = f'Delegated {stake_amount_norm} GRT to indexer {indexer}'
+                event.counterparty = CPT_THEGRAPH
+
+                # also account for the GRT burnt due to delegation tax
+                tokens_burnt = initial_amount_norm - stake_amount_norm
+                if tokens_burnt > 0:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:114]: Shouldn't this always be true? Or just going for some kidn of future proof thing where maybe the protocol removes the fee?

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+                        asset=A_GRT,
+                        balance=Balance(amount=tokens_burnt),
+                        location_label=delegator,
+                        notes=f'{tokens_burnt} GRT has been burnt as delegation tax',
+                        counterparty=CPT_THEGRAPH,
+                        address=context.tx_log.address,
+                    )
+                    return DecodingOutput(event=burn_event)
+                break
+        return DEFAULT_DECODING_OUTPUT
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:129]: This helps readability since it's the end of a loop and python does not have curlies

```suggestion
                break

        return DEFAULT_DECODING_OUTPUT
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+            tx_log=context.tx_log,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=A_GRT,
+            balance=Balance(),
+            location_label=delegator,
+            notes=(
+                f'Undelegated {tokens_amount_norm} GRT'
+                f' from indexer {indexer}'
+                f', lock expires in {lock_timeout_secs} seconds'
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:155]: Doesn't this fit in one or at least two lines? Then keep it so.
Also use present tense.

```suggestion
                f'Undelegate {tokens_amount_norm} GRT from '
                f'indexer {indexer}. Lock expires in {lock_timeout_secs} seconds'
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+                        notes=f'{tokens_burnt} GRT has been burnt as delegation tax',
+                        counterparty=CPT_THEGRAPH,
+                        address=context.tx_log.address,
+                    )
+                    return DecodingOutput(event=burn_event)
+                break
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_stake_locked(self, context: DecoderContext) -> DecodingOutput:
+        """Processes StakeDelegatedLocked event"""
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:132]: as I said not needed. Name is self-documenting

```suggestion
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+        if context.tx_log.topics[0] == TOPIC_STAKE_DELEGATED:
+            return self._decode_stake_delegated(context)
+        elif context.tx_log.topics[0] == TOPIC_STAKE_DELEGATED_LOCKED:
+            return self._decode_stake_locked(context)
+        elif context.tx_log.topics[0] == TOPIC_STAKE_DELEGATED_WITHDRAWN:
+            return self._decode_stake_withdrawn(context)
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_stake_delegated(self, context: DecoderContext) -> DecodingOutput:
+        """Processes StakeDelegated event"""
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:87]: docstring repeating the name is not needed

```suggestion
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+                f' from indexer {indexer}'
+                f', lock expires in {lock_timeout_secs} seconds'
+            ),
+            counterparty=CPT_THEGRAPH,
+            address=context.tx_log.address,
+        )
+        return DecodingOutput(event=event)
+
+    def _decode_stake_withdrawn(self, context: DecoderContext) -> DecodingOutput:
+        """Processes StakeDelegatedWithdrawn event"""
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:163]: docstring repeating the name is not needed

```suggestion
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+                event.location_label == delegator and
+                event.address == CONTRACT_THEGRAPH_STAKING and
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE
+            ):
+                initial_amount_norm = event.balance.amount
+                event.event_type = HistoryEventType.STAKING
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.balance = Balance(amount=stake_amount_norm)
+                event.notes = f'Delegated {stake_amount_norm} GRT to indexer {indexer}'
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:109]: Did not notice it before, but if you check most of our decoders we use present tense and start with a verb.

```suggestion
                event.notes = f'Delegate {stake_amount_norm} GRT to indexer {indexer}'
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+                if tokens_burnt > 0:
+                    burn_event = self.base.make_event_from_transaction(
+                        transaction=context.transaction,
+                        tx_log=context.tx_log,
+                        event_type=HistoryEventType.SPEND,
+                        event_subtype=HistoryEventSubType.FEE,
+                        asset=A_GRT,
+                        balance=Balance(amount=tokens_burnt),
+                        location_label=delegator,
+                        notes=f'{tokens_burnt} GRT has been burnt as delegation tax',
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:123]:
```suggestion
                        notes=f'Burn {tokens_burnt} GRT as delegation tax',
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+        # create action item that will modify the relevant Transfer event that will appear later
+        action_item = ActionItem(
+            action='transform',
+            from_event_type=HistoryEventType.RECEIVE,
+            from_event_subtype=HistoryEventSubType.NONE,
+            asset=A_GRT,
+            amount=tokens_amount_norm,
+            to_event_type=HistoryEventType.STAKING,
+            to_event_subtype=HistoryEventSubType.REMOVE_ASSET,
+            to_notes=f'Withdrew {tokens_amount_norm} GRT from indexer {indexer}',
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:181]: Present tense

```suggestion
            to_notes=f'Withdraw {tokens_amount_norm} GRT from indexer {indexer}',
```

File: rotkehlchen/tests/unit/decoders/test_main.py [hunk 23]
```diff
@@ -75,6 +75,7 @@ def test_decoders_initialization(ethereum_transaction_decoder: 'EthereumTransact
         'Safemultisig',
         'Stakedao',
         'Sushiswap',
+        'Thegraph',
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_main.py:78]: In the same file at line 92 you forgot to add the counterparty id. This is why the only single test failed

File: rotkehlchen/tests/unit/decoders/test_thegraph.py [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.STAKING,
+            event_subtype=HistoryEventSubType.DEPOSIT_ASSET,
+            asset=A_GRT,
+            balance=Balance(amount=FVal('998.98')),
+            location_label=ADDY_USER,
+            notes='Delegated 998.98 GRT to indexer 0x6125eA331851367716beE301ECDe7F38A7E429e7',
+            counterparty=CPT_THEGRAPH,
+            address=ADDY_THEGRAPH,
+        ), EvmEvent(
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_thegraph.py:54]: This approval event in between is annoying. Can you use `maybe_reshuffle_events` in the decoder to guarantee Delegate and burn come one after the other?

File: rotkehlchen/tests/unit/decoders/test_thegraph.py [hunk 25]
```diff
@@ -0,0 +1,175 @@
...
+            tx_hash=tx_hash,
+            sequence_index=297,
+            timestamp=timestamp,
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=A_GRT,
+            balance=Balance(),
+            location_label=ADDY_USER,
+            notes=(
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_thegraph.py:121]: try to keep in one line if possible or 2 if not. It's mostly about readability and splitting into multiple lines is hard to follow for big string.

**Review by LefterisJP — COMMENTED** [review]
Some smaller comments

**Review by LefterisJP — COMMENTED** [review]
Also the PR has 7 commits already. It should only have 1. Or 2. One for implementation and 1 for tests. As linked before in contribution guide we keep commits doing one thing and if multiple commits are for same reason combine them.

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py (hunk 14 above)
Comment (d347h-eth) [reply rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:114]: Should always be true, yes, so it's just a sanity-check to not send a negative amount into a new `EvmEvent` instance.

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+            counterparty=CPT_THEGRAPH,
+            address=context.tx_log.address,
+        )
+        return DecodingOutput(event=event)
+
+    def _decode_stake_withdrawn(self, context: DecoderContext) -> DecodingOutput:
+        delegator = hex_or_bytes_to_address(context.tx_log.topics[2])
+        if delegator is None or self.base.is_tracked(delegator) is False:
+            return DEFAULT_DECODING_OUTPUT
+        indexer = hex_or_bytes_to_address(context.tx_log.topics[1])
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:170]: Use newline after if. Helps readability

```suggestion

        indexer = hex_or_bytes_to_address(context.tx_log.topics[1])
```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+                    return DecodingOutput(event=burn_event)
+                break
+
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_stake_locked(self, context: DecoderContext) -> DecodingOutput:
+        delegator = hex_or_bytes_to_address(context.tx_log.topics[2])
+        if delegator is None or self.base.is_tracked(delegator) is False:
+            return DEFAULT_DECODING_OUTPUT
+        indexer = hex_or_bytes_to_address(context.tx_log.topics[1])
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:141]: Use newline after if. Helps readability

```suggestion
        indexer = hex_or_bytes_to_address(context.tx_log.topics[1])

```

File: rotkehlchen/chain/ethereum/modules/thegraph/decoder.py [hunk 28]
```diff
@@ -0,0 +1,187 @@
...
+        elif context.tx_log.topics[0] == TOPIC_STAKE_DELEGATED_WITHDRAWN:
+            return self._decode_stake_withdrawn(context)
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_stake_delegated(self, context: DecoderContext) -> DecodingOutput:
+        deposit_event, burn_event = None, None
+        delegator = hex_or_bytes_to_address(context.tx_log.topics[2])
+        if delegator is None or self.base.is_tracked(delegator) is False:
+            return DEFAULT_DECODING_OUTPUT
+        indexer = hex_or_bytes_to_address(context.tx_log.topics[1])
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/thegraph/decoder.py:91]: Use newline after if. Helps readability

```suggestion
        indexer = hex_or_bytes_to_address(context.tx_log.topics[1])

```

File: rotkehlchen/tests/unit/decoders/test_thegraph.py [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+            timestamp=timestamp,
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.STAKING,
+            event_subtype=HistoryEventSubType.REMOVE_ASSET,
+            asset=A_GRT,
+            balance=Balance(amount=FVal('1003.70342593701668535')),
+            location_label=ADDY_USER,
+            notes=(
+                'Withdraw 1003.70342593701668535 GRT'
+                ' from indexer 0x6125eA331851367716beE301ECDe7F38A7E429e7'
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_thegraph.py:168]: This is also a nitpick and can't enforce from linter but I have notices when you split a sentence string across multiple lines you end the line without a space and start the new with a space. Can you try to be doing the opposite? It's how we do it in the rest of the code.

Really minor nitpick to keep in mind. Here and in all other places you do it.

```suggestion
                'Withdraw 1003.70342593701668535 GRT '
                'from indexer 0x6125eA331851367716beE301ECDe7F38A7E429e7'
```

**Review by LefterisJP — APPROVED** [review]
Thanks. Looks good to me, will merge.

Do the small nitpicks in the next PR and keep them in mind. Also you would need to add a changelog in the next PR.
